### ISSUE TITLE
[P1] Blackwell GPUs excluded — need PyTorch cu126+ for SM12.0

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#17]
## Problem

Two RTX PRO 6000 Blackwell (96 GB each, SM 12.0) are excluded at runtime because the active PyTorch build (`cu124`, CUDA 12.4) does not ship SM 12.0 SASS for every op. The exclusion is logged on startup:

```
Note: Blackwell GPUs excluded (PyTorch cu118 < SM120). Upgrade to cu126+ for 5-GPU.
```

192 GB VRAM and ~600 TFLOPS BF16 sit idle. Training runs on 3 GPUs (H100 + 2×A6000) instead of 5.

---

## Root Cause

`setup_ags1.sh` installs `torch --index-url .../cu124`. PyTorch cu124 includes SM 12.0 PTX but not native SASS, so kernels JIT-recompile on first use and some paths (e.g. `fused_gqa_attention`) skip the `>= 1200` branch entirely on torch < 2.7.

---

## Fix

```bash
# In walking3 conda env:
pip install torch==2.7.0+cu126 torchvision torchaudio \
    --index-url https://download.pytorch.org/whl/cu126
```

Update `setup_ags1.sh` step 2 from `cu124` → `cu126` and pin `torch==2.7.0`.

---

## Affected Files

| File | Change needed |
|---|---|
| `setup_ags1.sh` | Change `cu124` → `cu126`, pin `torch==2.7.0` |
| `launch_7b_3gpu.sh` | Superseded by `launch_7b_5gpu.sh` (already exists) |
| `launch_7b_5gpu.sh` | Version guard already in place (`cuda >= 12.6` check); no change needed |
| `csrc/hetero_reduce/CMakeLists.txt` | Already handles `sm_120`; native SASS requires CUDA >= 12.8 toolkit (uses PTX fallback with cu126) |

---

## SM 12.0 Kernel Support Status (as of HEAD)

**hetero_reduce kernels** — all 10 `.cu` files in `csrc/hetero_reduce/` contain SM 12.0–aware logic (512-thread blocks, 4 CTAs/SM, cp.async paths). This work was completed in #71.

| Kernel | SM 12.0 status |
|---|---|
| `hetero_reduce.cu` | ✅ SM 12.0 dispatch + weight env var `HETERO_WEIGHT_SM120` |
| `hetero_ring_allreduce.cu` | ✅ double-buffered cp.async path for SM 12.0 |
| `fused_adam_heterogeneous.cu` | ✅ LR-scale env var `HETERO_ADAM_LR_SCALE_SM120`, 512-thread blocks |
| `fused_gradient_allreduce.cu` | ✅ 512-element tile for SM 12.0 |
| `pcie_adaptive_allreduce.cu` | ✅ cp.async.bulk / TMA path for SM 12.0 |
| `fused_swiglu_ln.cu` | ✅ 512-thread / 128-reg budget path |
| `fused_rope_hetero.cu` | ✅ 512-thread, 4 CTAs/SM |
| `cross_entropy_tp.cu` | ✅ 512-thread, 4 CTAs/SM |
| `tier_activation_offload.cu` | ✅ `__launch_bounds__(512, 4)` for SM 12.0 |
| `fused_layernorm_residual.cu` | ⚠️ No explicit SM 12.0 guard found; falls back to generic path |

**Other kernels:**
- `csrc/attention/fused_gqa_attention.cu`: has `#if __CUDA_ARCH__ >= 1200` block (line 273) ✅
- `csrc/adam/cpu_adam_tier.h`: `SM120` tier → `GPU_KERNEL` offload strategy ✅
- `flash-attn`: install skipped in `setup_ags1.sh` with note "may need manual build for SM 12.0" — needs verification ⚠️

**Native SASS vs PTX:** `csrc/hetero_reduce/CMakeLists.txt` enables `-arch=sm_120` native SASS only when CUDA toolkit >= 12.8. With cu126 (toolkit 12.6), Blackwell kernels compile via PTX forward-compat — correct but slower on first launch due to JIT. To get native SASS, upgrade toolkit to 12.8+ separately from the PyTorch wheel.

---

## Impact

3-GPU → 5-GPU. Estimated: 3,500 → 10,000+ tok/s throughput, training time ~54 days → ~18 days.

---

## Acceptance Criteria

- [ ] `setup_ags1.sh` installs `torch==2.7.0+cu126`
- [ ] All 5 GPUs detected and active at `launch_7b_5gpu.sh` startup (`CUDA_VISIBLE_DEVICES=0,1,2,3,4`)
- [ ] `torch.cuda.get_device_capability(0)` returns `(12, 0)` without error
- [ ] `csrc/hetero_reduce` kernels compile without error against cu126 toolkit
- [ ] `fused_layernorm_residual.cu`: confirm SM 12.0 path or add explicit guard
- [ ] Verify `flash-attn` builds for SM 12.0 or document workaround
- [ ] Smoke test: 10-step pretrain run on all 5 GPUs with loss decreasing

**Comment by dylanyunlon** [comment]
## 🔍 Claude-R3 审查报告 — Issue #17

**审查时间:** 2026-07-23  
**审查员:** Claude-R3 (PRD Review Automation)  
**结论: ⚠️ 条件通过 — 2 项缺陷需修复后合并**

---

### 准入标准逐项核查

#### ✅ Problem — 通过
- 有明确的运行时错误 log（来源已验证：`launch_7b_3gpu.sh` 第 111 行）
- 量化影响清晰：3-GPU → 5-GPU，192 GB VRAM 闲置
- **⚠️ 轻微不一致**：issue body 将当前安装描述为 `cu124`，但 log 消息写的是 `cu118`。代码库中 `setup_ags1.sh:23` 确实是 `cu124`，`launch_7b_3gpu.sh:111` 的 log 消息写的是 `cu118`（历史遗留）。两处版本号不一致，应统一说明。

#### ✅ Root Cause — 通过（需补行号）
- 核心原因正确：`setup_ags1.sh:23` 安装 `cu124`，PTX-only，部分算子 fallback
- `fused_gqa_attention.cu:273` 的 `#if __CUDA_ARCH__ >= 1200` 已验证存在 ✅
- **缺陷**：Root Cause 未给出 `setup_ags1.sh` 具体行号（应为第 23 行），仅说"step 2"。按准入标准需要代码行号引用。

#### ✅ Fix — 通过
- pip 命令完整，版本 pin 明确（`torch==2.7.0+cu126`）
- 变更目标文件和操作描述清晰

#### ⚠️ Affected Files — 部分通过
已验证文件存在性：

| 文件 | 验证状态 | 备注 |
|---|---|---|
| `setup_ags1.sh` | ✅ 存在，第 23 行需改 | 行号正确 |
| `launch_7b_3gpu.sh` | ✅ 存在 | 第 111 行 log 消息 cu118 为历史遗留 |
| `launch_7b_5gpu.sh` | ✅ 存在，第 22-37 行有版本 guard | 已验证 |
| `csrc/hetero_reduce/CMakeLists.txt` | ✅ 存在，第 60-67 行 CUDA >= 12.8 逻辑 | 已验证 |

**缺陷**：Files 表格未给出任何行号，仅描述变更类型。按准入标准应列出受影响行号范围。

#### ⚠️ Acceptance Criteria — 通过（有悬挂项）
- 7 条 AC 均可验证 ✅
- **悬挂风险**：AC #5（`fused_layernorm_residual.cu` SM 12.0 guard）和 AC #6（flash-attn SM 12.0 构建）已在 issue 内标注为 ⚠️，但尚无对应 fix PR/分支，这两项会阻塞 AC 验收。建议开子 issue 追踪或在本 issue Fix 节说明处理方案。

#### ✅ 无过时信息 — 通过
- 所提到的文件均存在于 HEAD
- Kernel 状态表与代码库一致（已逐一验证 10 个 `.cu` 文件）

---

### 代码库验证发现的额外问题（供修复时参考）

1. **`launch_7b_3gpu.sh:111` log 消息版本不一致**：log 写 `cu118`，实际 `setup_ags1.sh` 安装的是 `cu124`。完成 cu126 升级后两处均需更新，否则 log 消息会持续误导。

2. **`fused_layernorm_residual.cu` 无显式 SM 12.0 guard**：第 234 行使用 `LNResPolicy<SmVer>::kBlockSize` 模板，可能通过模板特化覆盖 SM 12.0，但 issue 标注的 ⚠️ 是正确的——需要补充确认或显式 guard。AC #5 直接对应此问题。

3. **`setup_ags1.sh:31` flash-attn**：`2>/dev/null || echo "...skipped..."` 会静默失败，不会阻塞脚本，但 SM 12.0 支持未验证。AC #6 需要明确的验收路径。

---

### 必须修复项（合并前）

- [ ] **Root Cause 节补充行号**：`setup_ags1.sh` 改为引用第 23 行（`pip install ... cu124`）
- [ ] **Affected Files 节补充行号**：各文件列出受影响行号范围（`setup_ags1.sh:23`、`launch_7b_3gpu.sh:111`）

### 建议修复项（不阻塞合并）

- [ ] 统一 Problem 节的 cu 版本描述（cu118 vs cu124 历史差异需一句话说明）
- [ ] AC #5/#6 悬挂项：开子 issue 或在本 issue 说明处理时间线
